### ISSUE TITLE
Screensharing in Firefox not working

### ISSUE THREAD

**Issue by foss-** [jitsi/jitsi-meet#1910]
macOS 10.12.6 + Firefox nightly 57.0a1 (2017-08-19) (64-bit)
also reproduced in Firefox 56b3

Reproduce
- create new jitsi room
- have mic + cam enabled
- start screensharing by pressing third button from the left

Firefox asks for permission to share the screen and offers to select a single window from a dropdown selection with the currently open windows.

Expected
Once that is allowed the screensharing should be up and running.

Currently
Error is shown

![permission denied](https://user-images.githubusercontent.com/5708172/29490470-ba592bb0-853d-11e7-946d-eef1d0d648c8.jpg)

Is this a known problem? I tried searching Firefox add-ons for "Jitsi Desktop sharing" but couldn't find any extension.

Maybe there is an existing bug for this? I couldn't find one.

**Comment by habi** [comment]
I have exactly the same problem with Firefox 55.0.1 (64-bit) on Ubuntu 16.04.

**Comment by saghul** [comment]
Hey there! I haven't been able to reproduce this. Are you able to do so on meet.jit.si? If you are using your own deployment, what do these options looks like in your config.js? https://github.com/jitsi/jitsi-meet/blob/master/config.js#L130-L145

**Comment by habi** [comment]
I’m having the issue on meet.jit.si

**Comment by saghul** [comment]
FTR, I tested this on Firefox 55 for macOS.

**Comment by foss-** [comment]
Hi @saghul thanks for looking into this. That's odd, so it's working as expected for you? I also tried 55.0.2 and it behaves the same as beta and nightly Firefox resulting in identical error message.

**Comment by saghul** [comment]
> That's odd, so it's working as expected for you?

Yes, it is. This is odd, I'm not sure what the cause could be!

**Comment by foss-** [comment]
@saghul if you want to see this in action in order to debug, feel free to ping.

**Comment by saghul** [comment]
@foss- I can reproduce this, but only if you select "No window" in the dropdown. If I select any window it starts to share it correctly. Our error handling seems incorrect here. Do you see the same?

**Comment by daimoc** [comment]
You can show the same error message if you click on "Don't allow" in screen sharing box. 
And if you had already click on "Don't allow", Firefox  temporary block screen sharing and next attempt to share screen open the popup with "Failed to install desktop sharing extension".

**Comment by foss-** [comment]
No longer seeing this with Firefox 57b4. Not sure why.

The remaining issue though I am seeing is, that when being asked which part to share, I am only able to share a single one of the open windows.

Is that expected behavior? And if yes, are there plans to allow for full screensharing? Should I open a new bug for that?

**Comment by thekyriarchy** [comment]
Firefox doesn't even show a screen sharing button for me, while chrome shows one it displays a permission denied error, unable to install extension. This is on meet.jitsi.net

**Comment by bgrozev** [comment]
> Firefox doesn't even show a screen sharing button for me, while chrome shows one it displays a permission denied error, unable to install extension. This is on meet.jitsi.net

Shreensharing has explicitly been disabled on firefox on meet.jitsi.net (I don't know why). Use meet.jit.si instead.

**Comment by thekyriarchy** [comment]
oh, that's so strange. should meet.jisti.net just forward to meet.jit.si or something? thank you for the help!

**Comment by parrenin** [comment]
Same problem here on meet.jit.si on debian 9 using the stock Firefox-ESR 52.
So, what is the solution?

**Comment by bgrozev** [comment]
I can sort of reproduce this if I try to select the "No window" option in the dropdown menu that firefox offers. Is this what you are seeing @parrenin , or do you see the error after selecting a valid window?

The "No window" option is the default one, and we should handle it better.

**Comment by parrenin** [comment]
@bgrozev: yes, the problem occurs when selection the "No window" option.
If I select a particular window, it works.

**Comment by bgrozev** [comment]
Closing this issue, since it is getting confusing. I've opened #2115 for the wrong error message problem. But I don't think this is high prio, and I don't even know if we can do anything about it with the API provided by firefox. Contributions would be welcome.

**Comment by foss-** [comment]
@bgrozev Not sure about closing this. But yes, it has become somewhat confusing.

Thanks for filing the new issue about the error message. #2115.

Another on-going problem with Firefox / Jitsi is that sharing of entire screen is not possible.

There are plenty of use cases where the best option is, to share the entire screen. It works in Chromium using the Jitsi Meeting extension.

Will that extension be brought to Firefox?

Searching for Jitsi on the Firefox add-on page brings a few exotic search results. Non of which is called "Jitsi Meetings" which seems to be the official (?) chromium extension. Also not a single add-on is compatible with Firefox 57+. So even if one of the add-ons did work, it will be incompatible with Firefox stable in about a week from now.

![search](https://user-images.githubusercontent.com/5708172/32133480-d5347f3a-bbd8-11e7-95f3-c37fcc9f5b89.png)

@bgrozev can you file the additional issues addressing the remaining problems (1. no firefox extension 2. no sharing of entire screen using firefox, the two are obviously related).

**Comment by parrenin** [comment]
BTW, the 'no window' label is bad IMHO.
I would rather write 'Full screen'.

**Comment by bgrozev** [comment]
@foss- the lack of extension is by design -- there is no need for an extension anymore. Since Firefox 52 screen sharing just works.

As for sharing a whole screen, the was a ticket and even some partial contributions, but I can not find it right now. In short, with Firefox the web application has to choose either individual windows or whole screens before it calls into the browser's window picker UI (while in chrome the browser's UI handles both). So we need some extra code and perhaps a second dialog box to ask the user which one is preferred.

**Comment by foss-** [comment]
That's all fine, but in Firefox 56 + 57b14 I do not see any option to share the entire screen. Is this https://github.com/jitsi/jitsi-meet/issues/496 ?

**Comment by saghul** [comment]
@foss- That's correct. We currently only support sharing a window, sharing the screen hasn't been implemented yet, look like that PR fell through the cracks :-(

That is however, not related to the need (or lack thereof) of an extension.

**Comment by tuxayo** [comment]
> look like that PR fell through the cracks

If someone has the time to push the two related PRs, this is the current status:
https://github.com/jitsi/lib-jitsi-meet/pull/129#issuecomment-324133198

**Comment by foss-** [comment]
As long as screensharing has not been implemented, should we keep this issue open?

**Comment by saghul** [comment]
@foss- I think this issue derailed a bit. There **is** screensharing support in Firefox. Feel free to open a new one about the lack of screen choosing though!

**Comment by foss-** [comment]
ok https://github.com/jitsi/jitsi-meet/issues/2240

**Comment by jtoy** [comment]
I noticed the Firefox extension works, but it doesnt give the option to show the whole desktop, just certain windows, in chrome you can choose your whole desktop, Is it supposed to work like that?

**Comment by bgrozev** [comment]
See #2240 . This is a limitation of the ff user interface, and it will be resolved once their changes land (I have not checked the progress there in a long time)

**Comment by bgrozev** [comment]
It's in the issue I liked to: https://bugzilla.mozilla.org/show_bug.cgi?id=1321221

I don't know whether it will just work automatically, or if we'll have to update jitsi-meet.

**Comment by daimoc** [comment]
This new getDisplayMedia method in Firefox is integrated to v66 (current Firefox nightly version).
I made a test with last Jitsi-Meet instance and we can select "Entire screen" for screen sharing in Firefox.
No modification needed.

**Comment by grahamglover** [comment]
Tried sharing screen today to a Firefox attendee and they did not get any prompt for said extension... nothing. Repeated it on a test meeting.  Mid meeting I had to switch to Zoom.

**Comment by saghul** [comment]
@grahamglover What Firefox version was that? I just gave it a quick test on meet.jit.si and it worked no problem. Notice that the extension is not needed anymore (it hasn't been for a long time). You should get a prompt from Firefox itself asking for what window / screen to share.

**Comment by grahamglover** [comment]
Hi @saghul - I was testing with a firefox browser in a remote desktop session.  I thought I was using the latest firefox, but must not have been as unable to replicate today. Will advise if I have the issue in the future, and will give Jitsi another shot for next web conference.

**Comment by siddhugolu** [comment]
I don't know whether this issue is getting tracked or not, but I can reproduce this on my Mac. Using Catalina 10.15.2 and Firefox nightly, when I select screen sharing - it brings the popup but selecting either of the two options does nothing.

**Comment by saghul** [comment]
@siddhugolu Please open a new issue. This one was opened ages ago and it's likely not related.

**Comment by Uffje** [comment]
https://support.mozilla.org/en-US/questions/1281482

Maybe this will help. You have to give permission from Catalina to firefox.

**Comment by siddhugolu** [comment]
> https://support.mozilla.org/en-US/questions/1281482
> 
> Maybe this will help. You have to give permission from Catalina to firefox.

Thanks, Uffje. I had forgotten about this issue as I seldom used Mac, but this fixes it. :)
